pkg/resiliency/resiliency.go

```diff
@@ -1,133 +1,535 @@
-// ------------------------------------------------------------
-// Copyright (c) Microsoft Corporation and Dapr Contributors.
-// Licensed under the MIT License.
-// ------------------------------------------------------------
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
 
 package resiliency
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"os"
 	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
 	"time"
 
 	grpc_retry "github.com/grpc-ecosystem/go-grpc-middleware/retry"
+	lru "github.com/hashicorp/golang-lru"
 	"gopkg.in/yaml.v2"
 
 	resiliency_v1alpha "github.com/dapr/dapr/pkg/apis/resiliency/v1alpha1"
 	operatorv1pb "github.com/dapr/dapr/pkg/proto/operator/v1"
 
+	"github.com/dapr/dapr/pkg/resiliency/breaker"
+	"github.com/dapr/kit/config"
 	"github.com/dapr/kit/logger"
+	"github.com/dapr/kit/retry"
 )
 
 const (
 	operatorRetryCount   = 100
 	operatorTimePerRetry = time.Second * 5
 )
 
-// Wrapper to let us define some helper functions without duplicating the domain definition.
-type Resiliency struct {
-	Resiliency resiliency_v1alpha.Resiliency
-}
+// ActorCircuitBreakerScope indicates the scope of the circuit breaker for an actor.
+type ActorCircuitBreakerScope int
 
-func LoadDefaultResiliency() *Resiliency {
-	return &Resiliency{resiliency_v1alpha.Resiliency{}}
-}
+const (
+	// ActorCircuitBreakerScopeType indicates the type scope (less granular).
+	ActorCircuitBreakerScopeType ActorCircuitBreakerScope = iota
+	// ActorCircuitBreakerScopeID indicates the type+id scope (more granular).
+	ActorCircuitBreakerScopeID
+	// ActorCircuitBreakerScopeBoth indicates both type and type+id are used for scope.
+	ActorCircuitBreakerScopeBoth // Usage is TODO
+)
+
+type (
+	// Provider is the interface for returning a `Runner` for the various
+	// resiliency scenarios in the runtime.
+	Provider interface {
+		// RoutePolicy returns the policy for a route.
+		RoutePolicy(ctx context.Context, name string) Runner
+		// EndpointPolicy returns the policy for a service endpoint.
+		EndpointPolicy(ctx context.Context, service string, endpoint string) Runner
+		// RemoveEndpoint removes an endpoint from a service.
+		// This would be used when a service is taken out of service.
+		RemoveEndpoint(name string, endpoint string)
```

Comment:
I think it'd be good to move this to an LRU like Actors and not have to worry about the additional method here. Just trying to keep concepts in the config lower where possible given how much it's already covering.

Reply:
You confirmed my question above. I agree and will make the change.